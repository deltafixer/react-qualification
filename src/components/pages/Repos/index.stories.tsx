import React from "react";
import Repos from "./";

export default {
  component: Repos,
  title: "Repos",
  // excludes dummy objects
  excludeStories: /.*Data$/,
};

// hardcoding styled div wrapper because storybooks theme is dark, app is white
// would make it smarter if it wasn't a test project (not to go deep in this when no need)

// since docs state that React components' state can't be modified in a story,
// not sure if the "Display Spinner or Data" should be described
export const Default = () => (
  <div style={{ background: "white", width: "100%", margin: "auto" }}>
    <Repos />
  </div>
);
