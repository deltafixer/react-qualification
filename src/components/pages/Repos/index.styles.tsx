import styled from "styled-components";

export const ReposStyleWrapper = styled.div`
  display: flex;
  flex-direction: column;
  font-size: 18px;

  > * {
    margin-bottom: 12px;
  }

  .number-of-results {
    font-weight: bold;
  }
`;
