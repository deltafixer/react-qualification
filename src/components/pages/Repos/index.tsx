import React, { useEffect, useState, useRef } from "react";
import { IGitHubRepo } from "../../../models/IGitHubRepo";
import { fetchMyRepos } from "../../../utils/api";
import { debounce } from "../../../utils/defaults";
import { Spinner, Input, List } from "../..";
import { ReposStyleWrapper } from "./index.styles";

const Repos = (): JSX.Element => {
  const [repos, setRepos] = useState([] as IGitHubRepo[]);
  const [searchString, setSearchString] = useState("");
  const [count, setCount] = useState(0);
  const loading = useRef(true);

  useEffect(() => {
    fetchMyRepos().then((fetchedRepos) => {
      loading.current = false;
      setRepos(fetchedRepos);
    });
  }, []);

  const searchInitiated = debounce((value: string) => {
    setCount(0);
    setSearchString(value);
  }, 400);

  return loading.current ? (
    <Spinner text={"Loading repositories..."} />
  ) : (
    <ReposStyleWrapper>
      <Input
        type="text"
        autoFocus
        placeholder="Type here to search for repos..."
        onChange={searchInitiated}
      />
      <List items={repos} itemFilter={searchString} resultCount={setCount} />
      {/* still playing with this div below :D */}
      <div className="number-of-results">Number of results:&nbsp;{count}</div>
    </ReposStyleWrapper>
  );
};

export default Repos;
