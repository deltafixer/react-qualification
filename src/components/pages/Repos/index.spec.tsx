import React from "react";
import { mount, ReactWrapper } from "enzyme";
import Repos from "./index";
import { act } from "react-dom/test-utils";

// mocking utils functions, trivial...
// debounce had to be cut off setTimeout due to Jest not waiting!
jest.mock("../../../utils/api", () => ({
  fetchMyRepos: () => {
    return Promise.resolve([]);
  },
}));

jest.mock("../../../utils/defaults", () => ({
  debounce: (func: any) => {
    return (...args: any) => {
      func(...args);
    };
  },
}));

jest.useFakeTimers();

describe("List component tests", () => {
  let wrapper: ReactWrapper;

  beforeAll(async () => {
    await act(async () => {
      wrapper = mount(<Repos />);
    });
  });

  afterAll(() => {
    wrapper.unmount();
  });

  it("renders correctly", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("runs debounce function when onChange event triggered via Input", () => {
    wrapper.update();
    wrapper.find("input").simulate("change", {
      target: {
        value: "some text",
      },
    });
  });
});
