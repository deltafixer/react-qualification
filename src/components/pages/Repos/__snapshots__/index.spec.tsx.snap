// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component tests renders correctly 1`] = `
"<Repos>
  <Spinner text=\\"Loading repositories...\\">
    <styled.div>
      <div className=\\"sc-AxirZ gMWsVi\\">
        <div className=\\"loader\\" />
        <div className=\\"loading\\">
          Loading repositories...
        </div>
      </div>
    </styled.div>
  </Spinner>
</Repos>"
`;
